### ISSUE TITLE
Reference file to big to build hash table

### ISSUE THREAD

**Issue by jlipovac** [Illumina/DRAGMAP#20]
Hi!

I'm trying to use DRAGMAP for mapping to bacteria database from kraken2 to compare the results I've got with minimap2 and bwa mem2. Database fna file has around 110GB. When I tried to build hash table from the reference fasta file I got message: 
Hash table generation failed: Padded reference sequence too long (114040334336); maximum 68719476736 supported
Error: boost::exception: /home/jlipovac/software/DRAGMAP/src/lib/workflow/GenHashTableWorkflow.cpp(315): Throw in function void dragenos::workflow::buildHashTable(const dragenos::options::DragenOsOptions&)
Dynamic exception type: boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::logic_error> >
std::exception::what: ERROR: Hash table generation failed - Padded reference sequence too long (114040334336); maximum 68719476736 supported

I'm just wondering if there is any way to build hash table for the whole fna file?

After getting this error I tried to filter the database file and left only subset of the bacteria on which I got mappings with minimap2 and bwa mem2. With this filtered database I built a hash table and then tried to align paired-end reads. Final result was that all the reads were unmapped which I find very strange because I got mappings with minimap2 and bwa mem2.
Do you have any thoughts on this? Am I doing something wrong?

**Comment by come-raczy** [comment]
No, at the moment it is not possible to build a hash table for such a large database of reference genomes.

The database construction replicates what is done on DRAGEN and has strong memory constraints. The maximum reference size would probably be around 64 billion bases, and even for this toy would need to fine tune several of the parameters for hash table construction. In particular, you would need to specify “ht-mem-limit=64GB".
Also, with large reference sequences, the hash table will not contain all the k-mers at every reference position, which will quickly lead to inefficiencies in mapping as DRAGMAP uses only every other seed on the read (this is currently hardcoded in DRAGMAP).
Finally, with the default seed length of 21 bases, chances are that there will be an increased probability of collisions in the hashtable and it might be necessary to increase it to 27 for better results.

Another issue that is likely to occur is that the database contains many different organisms, some of them being related. As a result, trying to map a read against that database might give unintended results. For instance, the mapper could report no mapping at all because the read maps to too many different organisms, or it could return a few mappings with MAPQ 0 because more than one organism contains the sequence. Altogether, DRAGMAP is missing one notion that is critical for metagenomics studies: the notion that the reference is in fact a collection of references from many organisms.
